src/Controller/ProductDetailsPageController.php

```
@@ -0,0 +1,47 @@
+<?php
+
+namespace Supermarket\Controller;
+
+use Exception;
+use Supermarket\Datastore\ProductRepository;
+use Supermarket\Renderer\Renderer;
+use Supermarket\Request;
+use Supermarket\Response;
+
+class ProductDetailsPageController implements PageController
+{
+    /**
+     * @var ProductRepository
+     */
+    private $productRepository;
+
+    /**
+     * @var Renderer
+     */
+    private $renderer;
+
+    public function __construct(ProductRepository $productRepository, Renderer $renderer)
+    {
+        $this->productRepository = $productRepository;
+        $this->renderer = $renderer;
+    }
+
+    public function execute(Request $request): Response
+    {
+        try {
+            $id = $request->get('id');
+            if ($id === null) {
+                throw new Exception('Invalid request.');
+            }
+
+            $productDetailsTemplate = './src/Template/ProductDetails.html';
+            $product = $this->productRepository->getProductById($id);
+            $content = $this->renderer->renderProductDetails($product, $productDetailsTemplate);
+            $response = new Response($content);
+        } catch (Exception $e) {
```

do you really wanna catch all exceptions and handle them in the same way? I don't think so. Best to catch the `ProductNotFoundException` separately and return 404 in that case (since the product does not exists) and catch `Exception` in a second catch where you can return e.g. 5xx error since is is an unknown/server issue (e.g. cannot connect to database).